### DISCUSSION TITLE
Centering on free_text()

### DISCUSSION THREAD

**Discussion by delta2echo** [py-pdf/pypdf#1722] [Ideas]
I apologize if this can already be done, but there does not seem to be a way to center justify text when creating annotations. 
It's possible to format the text in other ways, but no way to justify the text. 

Is this a feature that's planned for the future?

**Comment by MartinThoma** [comment]
> Is this a feature that's planned for the future?

That is an interesting question. In general, I think pypdf should (directly) support core-pdf features. And only those. That means if you can find it in the PDF specification, pypdf should support it. If it's not in there, pypdf should not support it. The one exception is text extraction.

I have seen `/TextAlign /Justify` in `Table 344 – Additional standard layout attributes specific to block-level structure elements`.

I guess you are talking about the FreeText annotation?

![image](https://user-images.githubusercontent.com/1658117/226171309-1ca8303e-f830-4ed1-8d29-d87472129dad.png)

Hence the following should center your annotation text:

```python
free_text_annotation[NameObject("/Q")] = NumberObject(1)
```

Feel free to add a PR to add this in `generics/_annotations.py::free_text` as a `text_align` parameter of type `Literal['left', 'center', 'right'] :-)

**Reply by delta2echo** [reply]
Assuming you mean this section (https://github.com/py-pdf/pypdf/blob/main/pypdf/generic/_annotations.py#L75-L144)
Something like this?:

```python
class AnnotationBuilder:
   ...
   @staticmethod
    def free_text(..., 
                  text_align:str="left"
                   )-> DictionaryObject:
    
    ... 
    
    free_text = DictionaryObject()
    
    ... 
    
   text_align =text_align.lower()
    if text_align in ("center", "right"):
           free_text[NameObject("/Q")] = NumberObject({"center":1, 
                                                       "right":2}[text_align])
           
    ....
    
    return free_text
```

**Reply by MartinThoma** [reply]
Yes, exactly!

I would make it a keyword-only parameters though. Hence:

```
   @staticmethod
    def free_text(..., 
                  *,  # <--- that makes all following parameters to be keyword-only
                  text_align:str="left"
                   )-> DictionaryObject:

**Reply by MartinThoma** [reply]
Also, please then extend the test: https://github.com/py-pdf/pypdf/blob/main/tests/test_generic.py#L708-L745

**Reply by delta2echo** [reply]
just for the sake of being complete, I think this is the full changes needed:

```
class AnnotationBuilder:
   ...
   @staticmethod
    def free_text(..., 
                  *,
                  text_align:str="left"
                   )-> DictionaryObject:
     
     text_align =text_align.lower()                                              #<----Bring this to the top. 
     if text_align not in ("left","center","right"):                             #<----Not sure if you want this but I would put it in.
          raise ValueError("Only: 'left', 'center', 'right' are acceptable text alignment values.")
          
    font_str = "font: "
    ... 
    font_str = f"{font_str};text-align:{text_align};color:#{font_color}"         #<--- This change is necessary as well
    ... 
    free_text = DictionaryObject()
    
    ... 
    if text_align in ("center", "right"):
           free_text[NameObject("/Q")] = NumberObject({"center":1, 
                                                       "right":2}[text_align])
           
    ....
    
    return free_text
```

**Comment by pubpub-zz** [comment]
@MartinThoma / @delta2echo 
use of an enum type might be better than str (easier to code using the hints)

**Reply by pubpub-zz** [reply]
also, shouldn't we wait for conclusion from #1741  ?

**Reply by MartinThoma** [reply]
Yes, we should probably wait for #1741 to come to a final conclusion.

**Reply by delta2echo** [reply]
When do you expect #1741 to be resolved? (just a rough guess is good)

**Reply by MartinThoma** [reply]
I guess **23rd of April**. I want to finish https://github.com/py-pdf/pypdf/pull/1745 with the current comments / missing parts this week. Then we have some time to discuss / iron out any issues. If / once we merge #1745, it will close #1741

**Comment by hansschl** [comment]
Hi all, my I ask whether the above has been abandonned? Would be sad since centering free text annotations is available in PDF and can be quite handy (and I would need it ;-) and there is already a concrete implementation proposal above that would do the job nicely!
(And I dont seem to find the above in the latest pypdf that I am using.)

**Reply by stefan6419846** [reply]
> Hi all, my I ask whether the above has been abandonned?

No.

> and there is already a concrete implementation proposal above that would do the job nicely!

Which seems to have never been converted into a PR by the author, thus it is not available in the releases.
